### PR TITLE
CRM-20652: Generalise changeFeeSelection function to be used for any entity

### PR DESCRIPTION
**Pull request by monishdeb**
* [CRM-20652: Generalise changeFeeSelection function to be used for any entity other](https://issues.civicrm.org/jira/browse/CRM-20652)

### CODE REVIEW

File: CRM/Price/BAO/LineItem.php [hunk 1]
```diff
@@ -605,4 +604,466 @@ public static function getLineItemArray(&$params, $entityId = NULL, $entityTable
...
+   */
+  public static function changeFeeSelections(
+    $params,
+    $entityID,
+    $entity,
+    $contributionId,
+    $feeBlock,
+    $lineItems,
+    $paidAmount
+  ) {
```
Comment (pradpnayak) [inline CRM/Price/BAO/LineItem.php]: I prefer not to change the sequence of parameters. I would have written as 
  public static function changeFeeSelections(
    $params,
    $entityID,
    $contributionId,
    $feeBlock,
    $lineItems,
    $paidAmount,
    $entity = 'participant'
  ) {

and would not change at rest of place.

**Comment by pradpnayak** [comment]
QA'd Partial Payment for Participant and Contribution. 

Financial Entries created are correct and also Contribution status.

File: CRM/Price/BAO/LineItem.php (hunk 1 above)
Comment (monishdeb) [inline CRM/Price/BAO/LineItem.php]: Reason behind this change is fn is called in only two places and I changed the parameter order as per the lineitem table column.

File: CRM/Price/BAO/LineItem.php [hunk 2]
```diff
@@ -992,6 +1038,40 @@ protected function _updateEntityRecordOnChangeFeeSelection($inputParams, $entity
...
+   * Helper function to retrieve financial trxn parameters to reverse
+   *  for given financial item identified by $financialItemID
+   *
+   * @param int $financialItemID
+   *
+   * @return array $financialTrxn
+   *
+   */
+  protected function _getRelatedCancelFinancialTrxn($financialItemID) {
+    $query = "SELECT ft.*
```
Comment (pradpnayak) [inline CRM/Price/BAO/LineItem.php]: Can we use api?

**Comment by colemanw** [comment]
@monishdeb looks like this needs to be rebased. Thanks.

**Comment by monishdeb** [comment]
@colemanw I have rebased the PR!

**Comment by colemanw** [comment]
@civicrm-builder retest this please

**Comment by monishdeb** [comment]
Closing in favor of #10370 as the commits are being cherry-picked.

**Pull request closed without merging**